docs/source/custom-collections.rst

```diff
@@ -0,0 +1,326 @@
+            self._dsk = dsk
+            self._keys = keys
+
+        def __dask_graph__(self):
+            return self._dsk
+
+        def __dask_keys__(self):
+            return self._keys
+
+        @staticmethod
+        def __dask_optimize__(dsk, keys, **kwargs):
+            # We cull unnecessary tasks here. Note that this isn't necessary,
+            # dask will do this automatically, this just shows one optimization
+            # you could do.
+            dsk2, _ = cull(dsk, keys)
+            return dsk2
+
+        # Use the threaded scheduler by default.
+        __dask_default_get__ = staticmethod(threaded_get)
+
+        def __dask_postcompute__(self):
+            # We want to return the results as a tuple, so our finalizer
+            # function is `tuple`. There are no extra arguments, so we also
+            # return an empty tuple.
+            return tuple, ()
+
+        def __dask_postpersist__(self):
+            # Since our __init__ takes a graph as its first argument, our
+            # rebuild function can just be the class itself. For extra
+            # arguments we also return a tuple containing just the keys.
+            return Tuple, (self._keys,)
+
+        def __dask_tokenize__(self):
+            # For tokenize to work we want to return a value that fully
+            # represents this object. In this case it's the list of keys
+            # to be computed.
+            return tuple(self._keys)
+
+Demonstrating this class:
+
+.. code:: python
+
+    >>> from dask_tuple import Tuple
+    >>> from operator import add, mul
+
+    # Define a dask graph
+    >>> dsk = {'a': 1,
+    ...        'b': 2,
+    ...        'c': (add, 'a', 'b'),
+    ...        'd': (mul, 'b', 2),
+    ...        'e': (add, 'b', 'c')}
+
+    # The output keys for this graph
+    >>> keys = ['b', 'c', 'd', 'e']
+
+    >>> x = Tuple(dsk, keys)
+
+    # Compute turns Tuple into a tuple
+    >>> x.compute()
+    (2, 3, 4, 5)
+
+    # Persist turns Tuple into a Tuple, with each task already computed
+    >>> x2 = x.persist()
+    >>> isinstance(x2, Tuple)
+    True
+    >>> x2.__dask_graph__()
+    {'b': 2,
+     'c': 3,
+     'd': 4,
+     'e': 5}
+    >>> x2.compute()
+    (2, 3, 4, 5)
+
+
+Checking if an object is a dask collection
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To check if an object is a dask collection, use ``dask.base.is_dask_collection``:
+
+.. code:: python
+
+    >>> from dask.base import is_dask_collection
+    >>> from dask import delayed
+
+    >>> x = delayed(sum)([1, 2, 3])
+    >>> is_dask_collection(x)
+    True
+    >>> is_dask_collection(1)
+    False
+
+
+.. _deterministic-hashing:
+
+Deterministic Hashing
+---------------------
+
+Dask implements its own deterministic hash function to generate keys based on
+the value of arguments. This function is available as ``dask.base.tokenize``.
+Many common types already have implementations of ``tokenize``, which can be
+found in ``dask/base.py``.
```

Comment:
Would it make sense for `tokenize` to be available from the `dask` namespace directly if it is being advertised as a public API function?

Reply:
I don't think `tokenize` should be top-level, as most users won't use it at all. It is public api though.